### PR TITLE
osc: update chapter marker positions when duration changes

### PR DESCRIPTION
**Pull request by Sagnac**
Commit 6abb7e3 updates the markers when the chapters change, but it
doesn't update their relative position at the bar when the duration
changes.

This means that adding chapters to a live stream would result in
corresponding chapter markers which were static while the duration
changed and thus their positions became incorrect over time until the
OSC was reinitialized.

This is fixed by observing the `duration` property if chapters are present
and reinitializing the OSC when the duration changes.

The `live_markers` user option, which determines whether the `duration`
property is observed when there are chapters, has been added in order to
allow disabling this behaviour as calling `request_init()` frequently
might have some impact on low-end systems.

The impact of `request_init()` on `render()` was measured to increase from
1-1.5 ms to 2-3 ms on a 2010 MacBook Air, while the impact was neglible
on a 2016 Surface Book (increasing only to an average of 1.4 ms from
1.3 ms for n=1500 render cycles).

The `live_markers` option is enabled by default.

### CODE REVIEW

**Comment by avih** [comment]
Not sure about the approach of this PR.

It unconditionally watches `duration` which changes very frequently for live streams, and is either wasteful if there are no chapters (which I think is the common case for live streams), or does really a lot of work very frequently with `request_init` if there are chapters.

I don't knot if that's possible with the current OSC code, but I'd think a better approach would be to re-calculate only the chapters if the conditions apply (we have chapters and duration changed) - similar to how updating the seek/position bar doesn't need an init of the whole OSC.

**Comment by Sagnac** [comment]
I'm not sure if there's an easy way around watching the duration property precisely because of how frequently it changes for live streams; the chapter indicators need to be updated every time the duration changes. I guess one strategy could be to include the update every render cycle as an alternative, but that would probably be more complicated, more frequent, and with limited benefit.

Right now this is practically no-op anyway if the duration is fixed or there are no chapters, so it only does work specifically when it needs to (duration changes + chapters present). Sure we still have to observe the duration when no chapters but I don't see an immediate drawback to this in the sense that it has any real impact.

The biggest concern is the use of `request_init` like you pointed out, we're calling this frequently and it does quite a few things (although the performance impact seems to be negligible).

> similar to how updating the seek/position bar doesn't need an init of the whole OSC.

From what I can see the current position slider is determined at `render_elements` and drawn at every render cycle, while the chapter indicators are drawn at `prepare_elements` which is why a simple `request_tick` doesn't work here. 

I guess a more direct way would be calling `prepare_elements` somewhere but I'm not sure we want to do that without initialising the OSC anyway. From the looks of it, the current code assumes the chapter positions are always going to remain static in relation to the playback percentage which is why the elements are drawn elsewhere and require a complete initialisation.

We might have to move a bit of code around if we want a more efficient strategy and at the very least this patch provides a fix without a chance of breaking anything. Any ideas?

**Comment by avih** [comment]
> Right now this is practically no-op anyway if the duration is fixed

True.

> or there are no chapters

Not true. It's wasteful in this case, because it's watched and called-back for nothing. It might not have a big performance impact on its own, especially with luajit, but it's still wasteful, and such things do add up.

> I'm not sure if there's an easy way around watching the duration property ... I guess one strategy could be to include the update every render cycle as an alternative, but that would probably be more complicated, more frequent, and with limited benefit.

I think this depends on how it's implemented. For starters, `duration` is not watched for live streams and yet I don't think there are missing updates due to this (ignoring chapters). It could be an oversight, or maybe other properties trigger a render frequently enough so that not observing `duration` doesn't become an issue.

What I'm trying to say here is that maybe explicitly watching `duration` is not needed, and instead maybe we can check at `render()` whether duration or chapters changed since the last render and act accordingly.

Did you test how frequently duration changes for live streams? Do you have some examples of live streams with chapters?

> From what I can see the current position slider is determined at `render_elements` and drawn at every render cycle, while the chapter indicators are drawn at `prepare_elements` which is why a simple `request_tick` doesn't work here.
> 
> I guess a more direct way would be calling `prepare_elements` ...

Yes, that's why I said that it might not be easy. Indeed the current design is such that the chapters are considered "static" and only calculated on init. I don't think calling `prepare_elements` instead of init would be a meaningful improvement, because that's the "heavy" part of init anyway, so I don't think it would save much.

The main thing I want to avoid is having an impact in use cases where it's not needed.

The second thing I prefer to avoid is the big impact of init when it is needed (we have chapters and duration changes). I've not decided yet whether it would be acceptable by default to init on every duration change when the conditions do apply.

Making this depend on an option is kind of dirty, because it's not really fair to ask the user to decide if they want this functionality at the cost of performance. But still, I think I could live with an option, because I assess that on low-end systems this could have a real usability impact (frequent inits).

The best approach, IMHO, would be that the chapters are taken out of the "static" part, and instead re-calculated as needed (only if chapters or duration changed since last time they were calculated) similar to how other parts are re-calculated on every render. This might take some effort though.

Bottom line, not sure yet what would be best/acceptable, but I think I could live with an option to enable this functionality.

**Comment by avih** [comment]
> Did you test how frequently duration changes for live streams?

So, I did test that with two streams. With one of them it seems to callback about once per second, and with the other stream it can be 2-3 times per second but inconsistently (few seconds without changes, then few changes in sub-second intervals).

I also tested how much impact `request_init` has on `render`, and on my low-ish-end system (high-end macbook air late 2010) it's about twice longer - from 1-1.5 ms to 2-3 ms. That's with lua 5.2 (not jit). I Imagine this would have an acceptable impact by default on most systems, but it could have a meaningful impact on real low end systems like ARM SOCs.

So here's what I suggest:
1. Keep the current code, including observing duration.
2. Keep it enabled by default, but add an option to disable the init on duration changes even if we do have chapters, because it can have meaningful impact on very low end systems. Don't forget to update the docs too.
3. It would be nice if you can make the option also disable the duration observation, but you'd need to ensure it's consistent if options change at runtime, and is not observed twice. I think it should be possible relatively cleanly.
4. At the commit message, explain the performance impact, with numbers (preferably with your own numbers, but you can add mine too - numbers from two systems are better than one).
5. At the commit message, replace the first paragraph with: "Commit 6abb7e3 updates the markers when the chapters change, but it doesn't update their relative position at the bar when the duration changes.".

After you do the above, we'll wait few days to allow others to comment.

**Comment by Sagnac** [comment]
> I don't think calling `prepare_elements` instead of init would be a meaningful improvement, because that's the "heavy" part of init anyway

> The best approach, IMHO, would be that the chapters are taken out of the "static" part, and instead re-calculated as needed (only if chapters or duration changed since last time they were calculated) similar to how other parts are re-calculated on every render. This might take some effort though.

I agree the best way would be refactoring in order to put the chapter elements into a separate function which could conditionally be called at `render`, but I'm not sure this is worth it considering how rare live streams with chapters are. 

Since probably few people use scripts which add chapters to live streams it makes sense to put this behind a user option instead. It keeps things simple and it doesn't run the risk of introducing bugs or possibly making things more complicated in order to address uncommon use cases.

Do you have a preference for the option name? I'm thinking either `updatemarkers` or `update_markers`.

> I also tested how much impact `request_init` has on `render`, and on my low-ish-end system (high-end macbook air late 2010) it's about twice longer - from 1-1.5 ms to 2-3 ms.

For n=1500 render cycles I got an average of ~1.3 ms without and ~1.4 ms with `request_init` on a ~2016 Surface Book.

> 3. It would be nice if you can make the option also disable the duration observation, but you'd need to ensure it's consistent if options change at runtime, and is not observed twice. I think it should be possible relatively cleanly.

This is easily done by checking if the option has changed at `update_options` (which is called every time any options change at runtime) and then enabling/disabling the observation appropriately. Something like:

```
if list["update_markers"] then
    if user_opts.update_markers then
        mp.observe_property("duration", "number", on_duration_change)
    else
        mp.unobserve_property(on_duration_change)
    end
end
```

The current code would then change to:

```
function on_duration_change()
    local chapters = mp.get_property_number("chapters", 0)
    if chapters > 0 then
        request_init()
    end
end

if user_opts.update_markers then
    mp.observe_property("duration", "number", on_duration_change)
end
```

If we choose not to gate the observation behind the user option and only want to gate the `request_init` call instead then we'd just need to change line 2659 in the current commit to:
```
if user_opts.update_markers and chapters > 0 then
```

Which way should we go about it?

**Comment by avih** [comment]
> This is easily done by checking if the option has changed at `update_options`

It can, but it's not a great idea, because it's out of the script's control, and generally it's intended to be informational rather than factual.

IMO, a better idea is to have it fully under script control, and also avoid the observation when it's not needed. Something like this? (untested):
```lua
-- request_init wrapped so that we can onobserve it on its own
function on_duration() request_init() end

local duration_watched = false;
function update_duration_watch()
    local want_watch = user_opts.live_chapters and
          (mp.get_property_number('chapter', 0) or 0) > 0 and
          true or false  -- ensure it's a boolean
    
    if (want_watch ~= duration_watched) then
        if want_watch then
            mp.observe_property("duration", nil, on_duration)
        else
            mp.unobserve_property(on_duration)
        end
        duration_watched = want_watch
    end
end

-- (avih) modify the existing chapter-list observer and update_options to call update_duration_watch

mp.observe_property("chapter-list", nil, function()
    update_duration_watch()
    request_init()
end)

function update_options(list)
    validate_user_opts()
    request_tick()
    visibility_mode(user_opts.visibility, true)
    update_duration_watch()
    request_init()
end
```

The name `live_chapters` is my own suggestion for this new option (I don't insist on it though, if you have better), and I think it can default to enabled.

**Comment by Sagnac** [comment]
> It can, but it's not a great idea, because it's out of the script's control

I suppose, but using `update_options` allows the solution to have simpler logic. It's up to you, both ways work.

> ```lua
> mp.get_property_number('chapter', 0) or 0
> ```

I think we want to get `chapters` instead of the current chapter here. I'm also unsure as to why we need two default values, as `nil` fallbacks to `0` with the first one anyway.

> ```lua
> true or false  -- ensure it's a boolean
> ```

I don't think this is necessary as the expression `(mp.get_property_number('chapter', 0) or 0) > 0` evaluates to either true or false and the user option is boolean

> ```lua
> --        Don't forget to call it also once on init after validate_user_opts()
> ```

The initial change notification from observing `chapter-list` calls `update_duration_watch()` on init so this might not be needed either.

> ```lua
> mp.observe_property("duration", nil, on_duration)
> ```

Out of curiosity, why should we observe `duration` as type `nil` instead of `number` here?

> The name `live_chapters` is my own suggestion for this new option (I don't insist on it though, if you have better)

How about `live_markers`? The chapter timestamps themselves are still accurate even without this option; what the option is addressing are the seek bar chapter indicators/markers.

Anyway, after modifying `want_watch` to:

```lua
local want_watch = user_opts.live_chapters and
                   mp.get_property_number('chapters', 0) > 0
```

I tested the code and it works, so I can go ahead and update the PR and we can go with this solution if you're happy with it. I still prefer using `update_options` since to me it's cleaner overall, but I like this solution too as we use a separate function to determine the observation and prevent polluting `update_options`.

**Comment by avih** [comment]
> I think we want to get `chapters` instead of the current chapter here

Yup. Typo.

> I'm also unsure as to why we need two default values, as `nil` falls back to `0` with the first one anyway.

That depends on the implementation, and I don't think `nil falls back to 0` - though I didn't look at the code. `nil` is generally a valid value (even if currently not for this property), so if for some reason it does return a valid `nil` then we want 0 for the numeric comparison, or else lua throws an exception. I believe that if a valid value/non-error nil is returned then it doesn't use the fallback. It's a safety measure.

> I don't think this is necessary as the expression `(mp.get_property_number('chapter', 0) or 0) > 0` evaluates to either true or false and the user option is boolean

Correct, but again, a safety measure, to make it independent of `user_opts.live_chapters` type. We need to ensure the value as boolean because the code below compares the values.

> The initial change notification from observing `chapter-list` calls `update_duration_watch()` on init so this might not be needed either.

Not needed indeed for the reason you described, but it is something we need to ensure is called on startup, so I prefer it explicit.

> Out of curiosity, why should we observe `duration` as type `nil` instead of `number` here?

Mainly because we don't use the value of the property here, so no need to fetch it.

> How about `live_markers`?

Sure. Sounds fine to me.

> I tested the code and it works

Nice. Thanks.

> I still prefer using update_options

I don't 100% understand what you mean by this. If you mean your original suggestion (of checking `list["live_markers"]` then it's not enough.

Functionally it's different than your original PR and also your second suggestion. The difference is that your solutions always (or depending on an option) observes duration, while my approach observes duration only when it's both enabled (option) _and_ we have chapters. So it can't be only at update_options, and whenever one of them changes we need to take both into account. So this approach is overall minimal code for the functionality.

Do feel free to suggest something else though.

> so I can go ahead and update the PR and we can go with this solution if you're happy with it

Sure, I'm good with it if you are, but again, if you prefer something else which takes what we discussed into account, do suggest.

**Comment by Sagnac** [comment]
> It's a safety measure.

I see, makes sense. I've left them in just in case then.

> while my approach observes duration only when it's both enabled (option) _and_ we have chapters

> So this approach is overall minimal code for the functionality.

Yes, I see now how your solution is the most parsimonious as `duration` is observed only when absolutely necessary. Although we could include checking for chapters if the option changes at `update_options` we'd just wind up with the same approach and it'd be messier, so it's better to put the observation decision in its own function.

I've updated the PR, let me know if I missed anything.

**Comment by avih** [comment]
>  ... considering how rare live streams with chapters are.

I asked before but without reply: do you have examples of such streams? are there known live streams with chapters? or is this purely for completeness sake? or for some other reason?

**Comment by Sagnac** [comment]
I can't recall ever seeing a live stream with chapters, so I'm not sure if there are any somewhere; I don't see why it wouldn't be possible though, but correct me if I'm wrong.

This should mostly address the problem from the perspective of using scripts which add chapters to live streams. The commit referenced here which added the re-init on `chapter-list` changes mentioned a SponsorBlock script which does so, and my own script which I use to save live streams also has the ability to add live chapters which is how I found the bug to begin with.

**Comment by avih** [comment]
Pushed as 96b246d9 with the following changes:
- Renamed `live_markers` to `livemarkers` - no-space convention as other options.
- Shortened the docs a bit.
- Modified one comment at `osc.lua`.

Thanks.

**Comment by avih** [comment]
By the way, this causes a regression where if there are chapters and duration changes (i.e. the very thing which this PR addresses), then the OSC buttons stop working or behave incorrectly.

That's because `osc_init()` which is used to re-render the chapter markers also resets/recreates all the OSC elements, but the mouse clicks implementation/handling depends on the elements staying unmodified between mouse-down and mouse-up, so `osc_init()` between mouse-down and mouse-up breaks mouse clicks.

I'll fix it soon.

**Comment by avih** [comment]
> By the way, this causes a regression

Fixed in 89efe820 .
